### PR TITLE
[CLR Profiler] Do not instrument on NuGet/Profiler version mismatch

### PR DESCRIPTION
**Pull request by zacharycmontoya**
Changes proposed in this pull request:
Modify the CLR Profiler so that we only mark the managed assembly as "loaded" when version of the managed assembly matches the version of the profiler. This fixes `TypeLoadException` issues (see #550) because the profiler will no longer replace method calls when other versions of the managed assembly are loaded. This is achieved by registering to the runtime's AssemblyLoad events so that on AssemblyLoadFinished we can read and compare the version of the managed assembly with the version of the profiler.

Additionally,  add a reproduction test project that's intended to reproduce this issue with AdoNet calls. This test cannot run in CI to reproduce the before/after scenario because the version of the repo (responsible for the version of the profiler) must be modified to a version that is earlier than the NuGet package dependency in the test project.

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 1]
```diff
@@ -150,6 +152,82 @@ CorProfiler::Initialize(IUnknown* cor_profiler_info_unknown) {
...
+  // Configure a version string to compare with the profiler version
+  WSTRINGSTREAM ws;
+  ws << ToWSTRING(assembly_metadata.version.major)
+      << '.'_W
+      << ToWSTRING(assembly_metadata.version.minor)
+      << '.'_W
+      << ToWSTRING(assembly_metadata.version.build);
+
+  // Check that Major.Minor.Build match the profiler version
+  if (ws.str() == ToWSTRING(PROFILER_VERSION)) {
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: This is the new assembly version check which, if successful, adds the AppDomainID to our internal `managed_profiler_loaded_app_domains` structure.

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Is there a better way to do this comparison besides building up a string?

File: src/Datadog.Trace.ClrProfiler.Native/clr_helpers.h [hunk 2]
```diff
@@ -230,7 +232,7 @@ struct AssemblyMetadata {
         assembly_token(assembly_token),
         version(Version(major, minor, build, revision)) {}
 
-  bool is_valid() const { return module_id != 0; }
+  bool IsValid() const { return module_id != 0; }
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Native/clr_helpers.h]: Random refactor to make the `IsValid` names consistent.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 1 above)
Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: What about comparing the `AssemblyMetadata` instance to a string? `PROFILER_VERSION` should always be a `major.minor.build` format, right?

File: reproductions/NugetProfilerVersionMismatch.TypeLoadException/Program.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql;
+
+namespace HttpMessageHandler.StackOverflowException
```
Comment (colin-higgins) [inline reproductions/NugetProfilerVersionMismatch.TypeLoadException/Program.cs]: Forget to update the namespace?

File: reproductions/NugetProfilerVersionMismatch.TypeLoadException/README.md [hunk 4]
```diff
@@ -0,0 +1 @@
+This program is a simple application that gets/sets values in a postgres database. To reproduce the actual TypeLoadException, the `tools/SynchronizeVersions` tool needs to be run to set the version of the live build to a version earlier than the version of the `Datadog.Trace.ClrProfiler.Managed` NuGet package specified in the .csproj file. When this program is run on .NET Core, .NET Core will load the newer version from the NuGet package instead of loading the correct version that lives alongside the profiler in the file system. Before making the proper version checks in the profiler, the ADO.NET methods will be instrumented and there will be entries in the profiler logs saying that method calls were replaced. After making the proper version checks in the profiler, we will enforce a strict version-match requirement and there will be no method replacements.
```
Comment (colin-higgins) [inline reproductions/NugetProfilerVersionMismatch.TypeLoadException/README.md]: Great explanation.
I bet we could figure out a way to run this in CI.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 5]
```diff
@@ -150,6 +152,82 @@ CorProfiler::Initialize(IUnknown* cor_profiler_info_unknown) {
...
+  if (!is_attached_) {
+    return S_OK;
+  }
+
+  if (debug_logging_enabled) {
+    Debug("AssemblyLoadFinished: ", assembly_id, " ", hr_status);
+  }
+
+  // keep this lock until we are done using the module,
+  // to prevent it from unloading while in use
```
Comment (colin-higgins) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Did you notice this happening?

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Nope, I just followed the convention of holding onto a lock while using the profiler like in the ModuleLoadFinished handler.

File: reproductions/NugetProfilerVersionMismatch.TypeLoadException/Program.cs (hunk 3 above)
Comment (zacharycmontoya) [reply reproductions/NugetProfilerVersionMismatch.TypeLoadException/Program.cs]: Good catch, fixed in c844ff61aafcbdc661212afdf6fa582a5c8b7cc8

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 1 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: The `AssemblyMetadata` class doesn't already have a ToString operation. I noticed that the `Version` struct we use does the same thing that I'm doing here (except that it does a four part assembly version) so this comparison should be fine as-is.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 5 above)
Comment (colin-higgins) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Any deadlock chances?

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: No. Only one of these callbacks is running at any point in time

**Pull request merged**